src/lib.rs

```diff
@@ -227,136 +299,176 @@ impl Mnemonic {
 	///
 	/// Some word lists don't guarantee that their words don't occur in other
 	/// word lists. In the extremely unlikely case that a word list can be
-	/// interpreted in multiple languages, an [Error::AmbiguousWordList] is
+	/// interpreted in multiple languages, an [Error::AmbiguousLanguages] is
 	/// returned, containing the possible languages.
-	pub fn language_of(s: &str) -> Result<Language, Error> {
-		// First we try wordlists that have guaranteed unique words.
-		let first_word = s.split_whitespace().next().unwrap();
-		if first_word.len() == 0 {
-			return Err(Error::BadWordCount(0));
-		}
-		for language in Language::all().iter().filter(|l| l.unique_words()) {
-			if language.find_word(first_word).is_some() {
-				return Ok(*language);
-			}
+	fn language_of<S: AsRef<str>>(mnemonic: S) -> Result<Language, Error> {
+		Mnemonic::language_of_iter(mnemonic.as_ref().split_whitespace())
+	}
+
+	/// Parse a mnemonic in normalized UTF8 in the given language.
+	pub fn parse_in_normalized(language: Language, s: &str) -> Result<Mnemonic, Error> {
+		let nb_words = s.split_whitespace().count();
+		if nb_words < MIN_NB_WORDS || nb_words % 6 != 0 || nb_words > MAX_NB_WORDS {
+			return Err(Error::BadWordCount(nb_words));
 		}
 
-		// If that didn't work, we start with all possible languages
-		// (those without unique words), and eliminate until there is
-		// just one left.
-		let mut langs: Vec<_> =
-			Language::all().iter().filter(|l| !l.unique_words()).cloned().collect();
-		for word in s.split_whitespace() {
-			langs.retain(|l| l.find_word(word).is_some());
+		// Here we will store the eventual words.
+		let mut words: [&'static str; MAX_NB_WORDS] = Default::default();
 
-			// If there is just one language left, return it.
-			if langs.len() == 1 {
-				return Ok(langs[0]);
+		// And here we keep track of the bits to calculate and validate the checksum.
+		// We only use `nb_words * 11` elements in this array.
+		let mut bits = [false; MAX_NB_WORDS * 11];
+
+		for (i, word) in s.split_whitespace().enumerate() {
+			let idx = language.find_word(word).ok_or(Error::UnknownWord(i))?;
+
+			words[i] = language.word_list()[idx];
+
+			for j in 0..11 {
+				bits[i * 11 + j] = idx >> (10 - j) & 1 == 1;
 			}
+		}
 
-			// If all languages were eliminated, it's an invalid word.
-			if langs.is_empty() {
-				return Err(Error::UnknownWord(word.to_owned()))
+		// Verify the checksum.
+		// We only use `nb_words / 3 * 4` elements in this array.
+		let mut entropy = [0u8; MAX_NB_WORDS / 3 * 4];
+		let nb_bytes_entropy = nb_words / 3 * 4;
+		for i in 0..nb_bytes_entropy {
```

Comment:
Hmm yeah that's how I took it from upstream and I also think it might be easier to reason about the code than to be setting bits inside the byte array. Not sure if the compiler optimizes this, probably not.